pysindy/optimizers/stlsq.py

```diff
@@ -125,28 +126,55 @@ def __init__(
         self.ridge_kw = ridge_kw
         self.initial_guess = initial_guess
         self.verbose = verbose
+        self.sparse_ind = sparse_ind
 
     def _sparse_coefficients(self, dim, ind, coef, threshold):
         """Perform thresholding of the weight vector(s)"""
         c = np.zeros(dim)
         c[ind] = coef
-        big_ind = np.abs(c) >= threshold
+        if self.sparse_ind is not None:
+            big_ind = ind
+            for i in range(dim):
+                if i in self.sparse_ind:
+                    big_ind[i] = np.abs(c[i]) >= threshold
+                else:
+                    big_ind[i] = True
+        if self.sparse_ind is None:
+            big_ind = np.abs(c) >= threshold
         c[~big_ind] = 0
         return c, big_ind
 
-    def _regress(self, x, y):
+    def _regress(self, x, y, dim):
         """Perform the ridge regression"""
         kw = self.ridge_kw or {}
-
-        with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", category=LinAlgWarning)
-            try:
-                coef = ridge_regression(x, y, self.alpha, **kw)
-            except LinAlgWarning:
-                # increase alpha until warning stops
-                self.alpha = 2 * self.alpha
-        self.iters += 1
-        return coef
+        sparse_ind = self.sparse_ind
+        if sparse_ind is None:
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=LinAlgWarning)
+                try:
+                    coef = ridge_regression(x, y, self.alpha, **kw)
+                except LinAlgWarning:
+                    # increase alpha until warning stops
+                    self.alpha = 2 * self.alpha
+            self.iters += 1
+            return coef
+        if sparse_ind is not None:
+            coef = np.zeros(dim)
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=LinAlgWarning)
+                try:
+                    coef_sparse = ridge_regression(x, y, self.alpha, **kw)
+                    coef_notsparse = ridge_regression(x, y, alpha=0, **kw)
```

Comment:
Don't use two separate regressions.  By way of explanation, consider the ridge regression objective for the case `sparse_ind=[1]`:

$$
\left\\|\left[\begin{matrix}1 & 1 \\\\ 0 & 1\end{matrix}\right] \left[\begin{matrix}x \\\\ y \end{matrix}\right] - \left[\begin{matrix}1 \\\\ 1\end{matrix}\right]\right\\|^2 + \left\\|\left[\begin{matrix}0 \\\\ y\end{matrix}\right]\right\\|^2.
$$

This is equivalent to the expression 

$$
(x+y-1)^2 + (y-1)^2 + y^2
$$

Your code instead finds $y$ from $(x+y-1)^2 + (y-1)^2 +x^2 + y^2$ and $x$ from $(x+y-1)^2 + (y-1)^2$ and concats them together, which isn't the same thing.  In the correct case, $x=y=.5$.  In the zero-penalty case, $x=0, y=1$.  In the full-penalty case, $0\lt x\lt.5,.5\lt y\lt1$.  your code would return $x=0, .5\lt y \lt 1$

The test only passes because our test has a diagonal matrix.  Instead, pass an array as the alpha parameter (See [the Ridge docs](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html)) and add the above equation as a new test.

Reply:
I understand why it does not work for non diagonal matrices.

However, when I try to pass an array as the alpha parameter, it does not work because:
While we are trying to minimize $\||Ax-b\||^2_2 + \alpha \||x\||^2_2$, the shape of A is (n_samples,n_features), shape of b is (n_samples, n_targets), shape of x is (n_targets, n_features). If $\alpha$ is to be passed as an array, according to the docs, its shape must be (n_targets,). In our case, n_targets is 1. 

I am unable to write alpha array as [0, 0, $\alpha$, $\alpha$] for our test case.

How do I resolve this?

Reply:
Oh good catch!  You might have to implement this yourself, motivated by below, and use linear regression

$$
\\|Ax-b\\|^2 + \alpha \\|x\\|^2 = \\left\\|\\left[\begin{matrix}A\\\\ \alpha I\end{matrix}\\right]x - \\left[\begin{matrix}b\\\\ 0\end{matrix}\\right]\\right\\|
$$

Does this make sense?

Reply:
I tried to do it using Linear Regression, however, both of the tests do not pass and the coef values are entirely different than what they should be. I joined A and $\alpha I$ matrices and added zeros to the b matrix, and performed linear regression on them. For the diagonal matrix test, the coefficients for nonsparse indices also changed and are not equal to one, as they should be. 

I tried a different implementation where I created an alpha-array 'A', which is a diagonal matrix $diag(0, 0, \alpha, \alpha)$ for sparse indices [2, 3].  and pass it in the equation for calcultion the minimum weight for ridge regression - 
$$w = (x^Tx + A)^{-1}x^Ty$$
This approach worked and the test with the nondiagonal matrix passed easily. For the one with the diagonal matrix, the test could only pass if I put max_iters = 1.
After each iteration, the regression calculations only take on the columns which have not been thresholded. So when a coefficient gets thresholded and the calculations move onto the next iteration, the column corresponding to that thresholded coefficient is not used in further computations. This means that when a coefficient is thresholded, I will also need to remove the corresponding index value and make changes to other indices in sparse_ind. Let me know if I should proceed with implementing this or if I should try to implement it using the Linear Regression equation you gave above.

Reply:
My comment about $\alpha I$ was just motivation - yes, I was looking for the $diag(0, 0, \alpha, \alpha)$.  Or more specifically, you only need the last two rows.  And - my mistake - these should be $\sqrt{\alpha}$ if directly appended to the feature matrix:

$$
\\left\\|\\left[\begin{matrix}A\\\\  \\left[\begin{matrix}0 & 0 & \sqrt{\alpha} & 0 \\\\ 0 & 0 & 0 & \sqrt{\alpha}\end{matrix} \\right]     \end{matrix}\\right]x - \\left[\begin{matrix}b\\\\ 0 \\\\ 0\end{matrix}\\right]\\right\\|^2
$$

This gives you normal equations of $(A^TA + diag(0, 0, \alpha, \alpha))^{-1}A^Tb$, or in your notation, $(x^T x + A)^{-1}x^Ty$ .  Let me know if that passes the test.

I would prefer that you manage to use scikitlearn for the actual regression.  $(x^T x + A)^{-1}x^Ty$ is correct, but practically, can be solved in ways that do not involve taking the (expensive) inverse.  Using an existing regression toolkit allows someone else to choose how to solve it - QR, pivoted LU, etc.

Reply:
There are 2 approaches I used for this: normal Linear Regression and Ridge Regression with $\alpha=0$. These two are supposed to be equivalent approaches but I get different coefficient arrays for both of them, the ridge regression with $\alpha=0$ being the one that passes all the tests. I do not understand why there is a difference in the coef array for the supposedly same approaches. I've pushed the updated code and tested it against many different inputs and it works for all of them, it's just with ridge regression with $\alpha=0$ and not normal Linear Regression.